torch/csrc/utils/future.h

```diff
@@ -128,7 +128,11 @@ class TORCH_API Future final {
     callbacks_.emplace_back(std::move(cb));
   }
 
-  private:
+  void addCallback(std::function<void(const Future<T>& future)> cb) {
```

Comment:
Doesn't have to be part of this PR, but does it make sense to pass a non-const reference to the callback such that we can move it around within the callback? This would avoid the other places where we pass the future in explicitly as a weak_ptr.

Reply:
btw, I wanted to think about this one (maybe we do)

`const Future<T>&` felt quite safe to do, while `Future<T>&` can be more dangerous (somewhat easier to sabotage oneself) with our pattern of allowing multiple callbacks to be added. 

For step 1, I wanted to address the simpler cases, while effectively "marking" the other cases for future consideration.

Reply:
Yes, it's true that it is definitely more dangerous. That was the reason in the initial PR I added the assert for use_count instead of moving out the value from the Future (which could break callbacks downstream).